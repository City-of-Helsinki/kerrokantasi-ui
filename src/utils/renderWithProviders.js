/* eslint react/prop-types: 0 */
import React from 'react';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import { LoginProvider } from 'hds-react';
import { IntlProvider } from 'react-intl';

import createStore from "../createStore";
import { userOidcConfig, apiTokenClientConfig } from './oidcConfig';

const loginProviderConfig = {
  userManagerSettings: userOidcConfig,
  apiTokensClientSettings: apiTokenClientConfig,
  sessionPollerSettings: { pollIntervalInMs: 10000 },
}

const renderWithProviders = (
  ui,
  { preloadedState = {}, locale = 'fi', history, store = createStore(preloadedState), ...renderOptions } = {},
) => {
  const Wrapper = ({ children }) => (
    <LoginProvider {...loginProviderConfig}>
      <IntlProvider locale={locale}>
        <Provider store={store}>{children}</Provider>
      </IntlProvider>
    </LoginProvider>
  );
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
};


export default renderWithProviders;
