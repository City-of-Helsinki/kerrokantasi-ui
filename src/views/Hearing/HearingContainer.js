import React, { lazy, Suspense } from 'react';
import Helmet from 'react-helmet';
import PropTypes from 'prop-types';
import isEmpty from 'lodash/isEmpty';
import { Switch, Route } from 'react-router-dom';
import { connect } from 'react-redux';
import { injectIntl, intlShape } from 'react-intl';

import * as HearingEditorSelector from '../../selectors/hearingEditor';
import Header from '../../components/Hearing/Header';
import LoadSpinner from '../../components/LoadSpinner';
import Section from '../../components/Hearing/Section/SectionContainer';
import getAttr from '../../utils/getAttr';
import { canEdit } from '../../utils/hearing';
import { fetchHearing as fetchHearingAction, setLanguage as setLanguageAction, fetchProjects } from '../../actions';
import { fetchHearingEditorMetaData } from '../../actions/hearingEditor';
import { getHearingWithSlug } from '../../selectors/hearing';
import { getUser } from '../../selectors/user';

const HearingEditor = lazy(() => import(/* webpackChunkName: "editor" */'../../components/admin/HearingEditor'));

export class HearingContainerComponent extends React.Component {
  constructor(props) {
    super(props);
    const {fetchProjectsList, fetchHearing, fetchEditorMetaData, match: {params}, location} = this.props;
    if (location.search.includes('?preview')) {
      // regex match to get the ?preview=key and substring to retrieve the key part
      fetchHearing(params.hearingSlug, location.search.match(/\?preview=([\w+-]+)/g)[0].substring(9));
    } else {
      fetchHearing(params.hearingSlug);
    }
    fetchEditorMetaData();
    fetchProjectsList();
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.location.state) {
      if (!isEmpty(nextProps.hearing)
        && nextProps.hearing.default_to_fullscreen
        && !nextProps.location.state.fromFullscreen) {
        this.props.history.push({
          pathname: `/${nextProps.hearing.slug}/fullscreen`,
          search: `?lang=${this.props.language}`
        });
      }
    } else if (!isEmpty(nextProps.hearing) && nextProps.hearing.default_to_fullscreen) {
      this.props.history.push({
        pathname: `/${nextProps.hearing.slug}/fullscreen`,
        search: `?lang=${this.props.language}`
      });
    }
    if (isEmpty(this.props.user) && isEmpty(this.props.hearing) && !isEmpty(nextProps.user)) {
      this.props.fetchHearing(nextProps.match.params.hearingSlug);
    }

    // re-render/fetch when navigate from a hearing container to another hearing container
    // since the component doesn't mount again
    if (nextProps.match.path === this.props.match.path
        && nextProps.location.pathname !== this.props.location.pathname
    ) {
      const { fetchHearing, fetchEditorMetaData, match: { params } } = nextProps;
      fetchHearing(params.hearingSlug);
      fetchEditorMetaData();
    }
  }

  render() {
    const {
      hearing,
      intl,
      language,
      user,
      labels,
      hearingDraft,
      hearingLanguages,
      isLoading,
      contactPersons,
      setLanguage
    } = this.props;

    return (
      <div className="hearing-page">
        {!isEmpty(hearing) ? (
          <React.Fragment>
            <Helmet
              title={getAttr(hearing.title, language)}
              meta={[{name: "description", content: getAttr(hearing.abstract, language)}]}
            />
            {(!isEmpty(user) && canEdit(user, hearing)) &&
              <Suspense fallback={<LoadSpinner />}>
                <HearingEditor
                  hearing={hearingDraft}
                  hearingLanguages={hearingLanguages}
                  labels={labels}
                  user={user}
                  isLoading={isLoading}
                  contactPersons={contactPersons}
                />
              </Suspense>
            }
            <div className="hearing-wrapper" id="hearing-wrapper">
              <Header
                hearing={hearing}
                language={language}
                intl={intl}
                setLanguage={setLanguage}
              />
              <Switch>
                <Route path="/:hearingSlug/:sectionId" component={Section} />
                <Route path="/:hearingSlug" component={Section} />
              </Switch>
            </div>
          </React.Fragment>
          ) : (
            <LoadSpinner />
          )
        }
      </div>
    );
  }
}

HearingContainerComponent.propTypes = {
  hearing: PropTypes.object,
  intl: intlShape.isRequired,
  language: PropTypes.string,
  match: PropTypes.object,
  user: PropTypes.object,
  labels: PropTypes.array,
  hearingDraft: PropTypes.object,
  hearingLanguages: PropTypes.array,
  isLoading: PropTypes.bool,
  contactPersons: PropTypes.array,
  fetchHearing: PropTypes.func,
  fetchEditorMetaData: PropTypes.func,
  setLanguage: PropTypes.func,
  history: PropTypes.object,
  location: PropTypes.object,
  fetchProjectsList: PropTypes.func
};

const mapStateToProps = (state, ownProps) => ({
  hearing: getHearingWithSlug(state, ownProps.match.params.hearingSlug),
  language: state.language,
  hearingDraft: HearingEditorSelector.getPopulatedHearing(state),
  hearingLanguages: state.hearingEditor.languages,
  labels: HearingEditorSelector.getLabels(state),
  user: getUser(state),
  isLoading: HearingEditorSelector.getIsLoading(state),
  contactPersons: HearingEditorSelector.getContactPersons(state)
});

const mapDispatchToProps = dispatch => ({
  fetchHearing: (hearingSlug, preview = false) => dispatch(fetchHearingAction(hearingSlug, preview)),
  fetchEditorMetaData: () => dispatch(fetchHearingEditorMetaData()),
  setLanguage: (lang) => dispatch(setLanguageAction(lang)),
  fetchProjectsList: () => dispatch(fetchProjects())
});

export default injectIntl(connect(mapStateToProps, mapDispatchToProps)(HearingContainerComponent));
