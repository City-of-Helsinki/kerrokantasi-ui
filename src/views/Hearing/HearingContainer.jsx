import React, { useCallback, useEffect, Suspense, lazy, useState, useMemo } from 'react';
import PropTypes from 'prop-types';
import { connect, useSelector } from 'react-redux';
import { Route, Routes, useLocation, useNavigate, useParams } from 'react-router-dom';
import { isEmpty } from 'lodash';
import Helmet from 'react-helmet';
import { useIntl } from 'react-intl';

import { organizationShape } from '../../types';
import { getHearingWithSlug } from '../../selectors/hearing';
import * as HearingEditorSelector from '../../selectors/hearingEditor';
import getUser from '../../selectors/user';
import { fetchHearing as fetchHearingAction, setLanguage as setLanguageAction, fetchProjects } from '../../actions';
import { fetchHearingEditorMetaData } from '../../actions/hearingEditor';
import getAttr from '../../utils/getAttr';
import { html2text } from '../../utils/commonUtils';
import { canEdit } from '../../utils/hearing';
import LoadSpinner from '../../components/LoadSpinner';
import Header from '../../components/Hearing/Header';
import Section from '../../components/Hearing/Section/SectionContainer';

const HearingEditor = lazy(() => import('../../components/admin/HearingEditor'));

const HearingContainerComponent = ({
  fetchProjectsList,
  fetchHearing,
  fetchEditorMetaData,
  hearingDraft,
  hearingLanguages,
  isLoading,
  labels,
  language,
  organizations,
  user,
}) => {
  const { hearingSlug } = useParams();
  const hearing = useSelector((state) => getHearingWithSlug(state, hearingSlug));
  const location = useLocation();
  const navigate = useNavigate();
  const intl = useIntl();
  const [shouldFetch, setShouldFetch] = useState(() => hearingSlug && isEmpty(hearing));

  const userCanEditHearing = useMemo(() => canEdit(user, hearing), [hearing, user]);

  const fetchHearingData = useCallback(() => {
    if (hearingSlug !== null && shouldFetch) {
      if (location.search.includes('?preview')) {
        // regex match to get the ?preview=key and substring to retrieve the key part
        fetchHearing(hearingSlug, location.search.match(/\?preview=([\w+-]+)/g)[0].substring(9)).then(() => {
          setShouldFetch(false);
        });
      } else {
        fetchHearing(hearingSlug).then(() => {
          setShouldFetch(false);
        });
      }
      fetchProjectsList();
    }
  }, [fetchHearing, fetchProjectsList, location.search, hearingSlug, shouldFetch]);

  useEffect(() => {
    if (isEmpty(hearing) && !isLoading) {
      fetchHearingData();
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [isLoading]);

  useEffect(() => {
    if (!isEmpty(hearing) && !isEmpty(user) && userCanEditHearing) {
      fetchEditorMetaData();
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [userCanEditHearing]);

  useEffect(() => {
    if (location.state) {
      if (!isEmpty(hearing) && hearing.default_to_fullscreen && !location.state.fromFullscreen) {
        navigate({
          pathname: `/${hearing.slug}/fullscreen`,
          search: `?lang=${language}`,
        });
      }
    } else if (!isEmpty(hearing) && hearing.default_to_fullscreen) {
      navigate({
        pathname: `/${hearing.slug}/fullscreen`,
        search: `?lang=${language}`,
      });
    }
  }, [hearing, language, location.state, navigate]);

  const helmetMeta = [
    { name: 'description', content: html2text(getAttr(hearing.abstract, language)) },
    { property: 'og:description', content: html2text(getAttr(hearing.abstract, language)) },
  ];

  if (hearing?.main_image?.url) {
    helmetMeta.push({ property: 'og:image', content: hearing.main_image.url });
  }

  return (
    <div className='hearing-page'>
      {!isEmpty(hearing) ? (
        <>
          <Helmet title={getAttr(hearing.title, language)} meta={helmetMeta} />
          {!isEmpty(user) && canEdit(user, hearing) && (
            <Suspense fallback={<LoadSpinner />}>
              <HearingEditor
                data-testid='hearingEditor'
                hearing={hearingDraft}
                hearingLanguages={hearingLanguages}
                labels={labels}
                user={user}
                isLoading={isLoading}
                organizations={organizations}
                navigate={navigate}
              />
            </Suspense>
          )}
          <div className='hearing-wrapper' id='hearing-wrapper'>
            <Header hearing={hearing} language={language} />
            <Routes>
              <Route path='/:sectionId' element={<Section />} />
              <Route path='/' element={<Section />} />
            </Routes>
          </div>
        </>
      ) : (
        (shouldFetch && <LoadSpinner />) || (
          <h1 style={{ textAlign: 'center' }}>{intl.formatMessage({ id: 'hearingNotFound' })}</h1>
        )
      )}
    </div>
  );
};

HearingContainerComponent.propTypes = {
  language: PropTypes.string,
  user: PropTypes.object,
  labels: PropTypes.array,
  hearingDraft: PropTypes.object,
  hearingLanguages: PropTypes.array,
  isLoading: PropTypes.bool,
  organizations: PropTypes.arrayOf(organizationShape),
  fetchHearing: PropTypes.func,
  fetchEditorMetaData: PropTypes.func,
  setLanguage: PropTypes.func,
  fetchProjectsList: PropTypes.func,
};

const mapStateToProps = (state) => ({
  language: state.language,
  hearingDraft: HearingEditorSelector.getPopulatedHearing(state),
  hearingLanguages: state.hearingEditor.languages,
  labels: HearingEditorSelector.getLabels(state),
  user: getUser(state),
  isLoading: HearingEditorSelector.getIsLoading(state),
  organizations: state.hearingEditor.organizations.all,
});

const mapDispatchToProps = (dispatch) => ({
  fetchHearing: (hearingSlug, preview = false) => dispatch(fetchHearingAction(hearingSlug, preview)),
  fetchEditorMetaData: () => dispatch(fetchHearingEditorMetaData()),
  setLanguage: (lang) => dispatch(setLanguageAction(lang)),
  fetchProjectsList: () => dispatch(fetchProjects()),
});

export default connect(mapStateToProps, mapDispatchToProps)(HearingContainerComponent);
