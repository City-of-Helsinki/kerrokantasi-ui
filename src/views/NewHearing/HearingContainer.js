import React from 'react';
import {connect} from 'react-redux';
import {Switch, Route, Redirect} from 'react-router-dom';
import Section from '../../components/Hearing/Section/SectionContainer';
import Header from '../../components/Hearing/Header';
import WrappedCarousel from '../../components/Carousel';
import {getHearingWithSlug} from '../../selectors/hearing';
import PropTypes from 'prop-types';
import {fetchHearing as fetchHearingAction} from '../../actions';
import LoadSpinner from '../../components/LoadSpinner';
import isEmpty from 'lodash/isEmpty';
import { injectIntl, intlShape } from 'react-intl';
import {SectionTypes} from '../../utils/section';
import {canEdit} from '../../utils/hearing';
import HearingEditor from '../../components/admin/HearingEditor';
import * as HearingEditorSelector from '../../selectors/hearingEditor';
import { initNewHearing, fetchHearingEditorMetaData } from '../../actions/hearingEditor';

export class HearingContainer extends React.Component {
  componentWillMount() {
    const {fetchHearing, fetchEditorMetaData, match: {params}} = this.props;
    fetchHearing(params.hearingSlug);
    fetchEditorMetaData();
  }

  render() {
    const {
      hearing,
      intl,
      language,
      match,
      user,
      labels,
      hearingDraft,
      hearingLanguages,
      isLoading,
      contactPersons
    } = this.props;
    const mainSectionId = !isEmpty(hearing) ? hearing.sections.find(section => section.type === SectionTypes.MAIN).id : null;

    return (
      <div className="hearing-page">

        {!isEmpty(hearing) ?
          <div>
            {(!isEmpty(user) && canEdit(user, hearing)) &&
              <HearingEditor
                hearing={hearingDraft}
                hearingLanguages={hearingLanguages}
                labels={labels}
                user={user}
                isLoading={isLoading}
                contactPersons={contactPersons}
              />
            }
            <div className="hearing-wrapper" id="hearing-wrapper">
              <Header hearing={hearing} activeLanguage={language} intl={intl}/>
              <WrappedCarousel hearing={hearing} intl={intl} language={language}/>
              <Switch>
                <Route path="/:hearingSlug/:sectionId" component={Section} />
                <Redirect from="/:hearingSlug" to={`/${match.params.hearingSlug}/${mainSectionId}`} />
              </Switch>
            </div>
          </div>
        : <LoadSpinner />
      }
      </div>

    );
  }
}

HearingContainer.propTypes = {
  hearing: PropTypes.object,
  intl: intlShape.isRequired,
  language: PropTypes.string,
  match: PropTypes.object
};

const mapStateToProps = (state, ownProps) => ({
  hearing: getHearingWithSlug(state, ownProps.match.params.hearingSlug),
  language: state.language,
  hearingDraft: HearingEditorSelector.getPopulatedHearing(state),
  hearingLanguages: state.hearingEditor.languages,
  labels: HearingEditorSelector.getLabels(state),
  user: state.user,
  isLoading: HearingEditorSelector.getIsLoading(state),
  contactPersons: HearingEditorSelector.getContactPersons(state)
});

const mapDispatchToProps = dispatch => ({
  fetchHearing: (hearingSlug, preview = false) => dispatch(fetchHearingAction(hearingSlug, preview)),
  fetchEditorMetaData: () => dispatch(fetchHearingEditorMetaData())
});

export default injectIntl(connect(mapStateToProps, mapDispatchToProps)(HearingContainer));
