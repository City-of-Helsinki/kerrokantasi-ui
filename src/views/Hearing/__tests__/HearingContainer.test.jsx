import React from 'react';
import { MemoryRouter } from 'react-router-dom';

import { HearingContainerComponent } from '../HearingContainer';
import { mockStore, getIntlAsProp } from '../../../../test-utils';
import renderWithProviders from '../../../utils/renderWithProviders';

const renderComponent = (propOverrides) => {
  const {
    labels,
    sectionComments,
    hearingLists: { allHearings },
    mockHearingWithSections,
    user,
  } = mockStore;
  const props = {
    labels: labels.data,
    hearing: mockHearingWithSections.data,
    hearingDraft: {},
    match: {
      params: {
        hearingSlug: allHearings.data[0].slug,
      },
    },
    location: {
      pathname: `/${allHearings.data[0].slug}`,
      search: '',
    },
    sectionComments,
    fetchHearing: () => null,
    language: 'fi',
    user,
    isLoading: false,
    fetchEditorMetaData: () => null,
    fetchProjectsList: () => null,
    ...propOverrides,
  };

  return renderWithProviders(
    <MemoryRouter>
      <HearingContainerComponent intl={getIntlAsProp()} {...props} />
    </MemoryRouter>,
  );
};

describe('<HearingContainer />', () => {
  it('should render correctly', () => {
    const { container } = renderComponent();

    expect(container).toMatchSnapshot();
  });
});
