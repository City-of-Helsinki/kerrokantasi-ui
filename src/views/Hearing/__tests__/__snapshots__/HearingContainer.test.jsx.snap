// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HearingContainer /> should render correctly 1`] = `
<div>
  <div
    class="hearing-page"
  >
    <div
      class="hearing-wrapper"
      id="hearing-wrapper"
    >
      <div
        class="header-section"
      >
        <div
          class="container"
        >
          <div
            class="hearing-header"
          >
            <div
              class="row"
            >
              <div
                class="col-md-9"
              >
                <h1
                  class="hearing-header-title"
                >
                  Ideoi Isosaaren tulevaisuutta
                </h1>
              </div>
            </div>
            <a
              href="/undefined?headless=false"
            >
              <span
                class="fa fa-arrow-left"
              />
               
              <span>
                backToHearingMain
              </span>
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
